tests/test_cli.py

```diff
@@ -558,8 +576,100 @@ def test_cli_data_types_add_columns(self, strict_writer, all_db_checkpoint_manag
+            )
+        ]
+    })
+    return [p1, p2][page]
+
+
+@pytest.mark.dbtest
+class TestCLIPaginationMode(object):
+    def test_cli_pagination_fresh(self, writer, all_db_checkpoint_manager):
+        checkpoint_manager = all_db_checkpoint_manager.for_dataset("case", ["Case"])
+
+        _pull_mock_data(writer, all_db_checkpoint_manager, get_indexed_on_client(0), 'tests/013_ConflictingTypes.xlsx')
+        self._check_data(writer, [["doc 1"], ["doc 2"]], "Case")
+        self._check_checkpoint(checkpoint_manager, '2012-04-24T05:13:01', 'doc 2')
+
+        _pull_mock_data(writer, all_db_checkpoint_manager, get_indexed_on_client(1), 'tests/013_ConflictingTypes.xlsx')
+        self._check_data(writer, [["doc 1"], ["doc 2"], ["doc 3"], ["doc 4"]], "Case")
+        self._check_checkpoint(checkpoint_manager, '2012-04-26T05:13:01', 'doc 4')
+
+    def test_cli_pagination_legacy(self, writer, all_db_checkpoint_manager):
+        """Test that we continue with the same pagination mode as was already in use"""
+
+        checkpoint_manager = all_db_checkpoint_manager.for_dataset("case", ["Case"])
+        # simulate previous run with legacy pagination mode
+        checkpoint_manager.set_checkpoint('2012-04-24T05:13:01', PaginationMode.date_modified, is_final=True)
+
+        client = MockCheckpointingClient({
+            'case': [
+                (
+                    {'limit': DEFAULT_BATCH_SIZE, 'order_by': 'server_date_modified', 'server_date_modified_start': '2012-04-24T05:13:01'},
+                    [
+                        {'id': "doc 1", 'name': 'n1', 'server_date_modified': '2012-04-25T05:13:01.000000Z'},
+                        {'id': "doc 2", 'name': 'n2', 'server_date_modified': '2012-04-26T05:13:01.000000Z'}
+                    ]
+                )
+            ]
+        })
+
+        _pull_mock_data(writer, all_db_checkpoint_manager, client, 'tests/013_ConflictingTypes.xlsx')
+        self._check_data(writer, [["doc 1"], ["doc 2"]], "Case")
+        self._check_checkpoint(checkpoint_manager, '2012-04-26T05:13:01', 'doc 2', PaginationMode.date_modified.name)
+
+    def test_cli_pagination_start_over(self, writer, all_db_checkpoint_manager):
+        """Test that we switch to the new pagination mode when using 'start_over'"""
+        checkpoint_manager = all_db_checkpoint_manager.for_dataset("case", ["Case"])
+        # simulate previous run with legacy pagination mode
+        checkpoint_manager.set_checkpoint('2012-04-24T05:13:01', PaginationMode.date_modified, is_final=True)
+
+        _pull_mock_data(writer, all_db_checkpoint_manager, get_indexed_on_client(0), 'tests/013_ConflictingTypes.xlsx', start_over=True)
+        self._check_data(writer, [["doc 1"], ["doc 2"]], "Case")
+        self._check_checkpoint(checkpoint_manager, '2012-04-24T05:13:01', 'doc 2')
+
+    def test_cli_pagination_since(self, writer, all_db_checkpoint_manager):
+        """Test that we use to the new pagination mode when using 'since'"""
```

Comment:
this confused for a moment since it seems odd that you'd ever have a checkpoint and then manually override it. what's the use case for that?

Reply:
not sure if there is a use case but it's not disallowed so just wanted to test it. I also realized now that if you pass in `--since` or `--until` then we don't do checkpointing at all. I'm going to update the test to reflect that.

Reply:
8f57ed9